src/AspNet.Security.OAuth.DingTalk/DingTalkAuthenticationHandler.cs

```diff
@@ -0,0 +1,153 @@
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace AspNet.Security.OAuth.DingTalk
+{
+    public class DingTalkAuthenticationHandler : OAuthHandler<DingTalkAuthenticationOptions>
+    {
+        public DingTalkAuthenticationHandler(
+            [NotNull] IOptionsMonitor<DingTalkAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
+        {
+            string stateValue = Options.StateDataFormat.Protect(properties);
+            var address = QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, new Dictionary<string, string>
+            {
+                ["appid"] = Options.ClientId,
+                ["response_type"] = "code",
+                ["scope"] = FormatScope(),
+                ["redirect_uri"] = redirectUri,
+                ["state"] = stateValue
+            });
+            return address;
+        }
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+        {
+            using var payload = JsonDocument.Parse("{\"access_token\":\"" + context.Code + "\"}");
+            return await Task.FromResult(OAuthTokenResponse.Success(payload));
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(
+            [NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] OAuthTokenResponse tokens)
+        {
+            var timestamp = GetTimeStamp();
+
+            var address = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, new Dictionary<string, string>
+            {
+                ["accessKey"] = Options.ClientId,
+                ["timestamp"] = timestamp,
+                ["signature"] = Signature(timestamp, Options.ClientSecret),
+            });
+
+            using var request = new HttpRequestMessage(HttpMethod.Post, address);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Content = new StringContent("{\"tmp_auth_code\":\"" + tokens.AccessToken + "\"}", System.Text.Encoding.UTF8, "application/json");
```

Comment:
As above, don't use string formatting to create JSON payloads.

Reply:
Has been adjusted